Upgrade to Polkadot 2412

drgora opened the pull request:
This PR upgrades Polkadot dependencies to version 2412-5.

To limit the size of the PR, some changes are left for future updates. These include:
- re-run the benchmarks to update weights
- (optional) enable approval-voting-parallel in the relay chain node (https://github.com/paritytech/polkadot-sdk/pull/4849)

la10736 commented on runtime/src/lib.rs:
```diff
@@ -1804,7 +1864,7 @@ impl_runtime_apis! {
         fn dispatch_benchmark(
             config: frame_benchmarking::BenchmarkConfig
-        ) -> Result<Vec<frame_benchmarking::BenchmarkBatch>, sp_runtime::RuntimeString> {
+        ) -> Result<Vec<frame_benchmarking::BenchmarkBatch>, alloc::string::String> {
```
That is something that maybe we missed before: we can works directly wit `alloc` and `core` instead of using `sp_std` crate to access. If it's true is better to try to fix it around all the codebase.

la10736 reviewed (approved):
In general LGTM. Just two minor note if you want to address them.

DanieleDiBenedetto reviewed (commented):
On the failing zombienet test... might it be related to the introduction of the new GeneralExtension trait, that now pallet_transaction_payment implements, thus adding a non-negligible cost now ?

la10736 commented:
> On the failing zombienet test... might it be related to the introduction of the new GeneralExtension trait, that now pallet_transaction_payment implements, thus adding a non-negligible cost now ?

I'm not betting about it. I did the math and we are quite far from it: the weight associated to the new `pallet_transaction_payment()` is since 1 * 10^8  but our error is in the order of 10^18.

Maybe we should try to rerun the verifier benchmarks ?

DanieleDiBenedetto commented:
> > On the failing zombienet test... might it be related to the introduction of the new GeneralExtension trait, that now pallet_transaction_payment implements, thus adding a non-negligible cost now ?
> 
> I'm not betting about it. I did the math and we are quite far from it: the weight associated to the new `pallet_transaction_payment()` is since 1 * 10^8 but our error is in the order of 10^18.
> 
> Maybe we should try to rerun the verifier benchmarks ?

But it seems it's not that the "happy path verifier cost" has decreased, but the cost of checking whether the verifier is disabled or not has increased significantly (which is strange considering that in this case the extrinsic does nothing)

drgora commented:
> On the failing zombienet test... might it be related to the introduction of the new GeneralExtension trait, that now pallet_transaction_payment implements, thus adding a non-negligible cost now ?

Turns out this is related to the new `TxExtension`. It seems that now the cost for each single extension is factored out from the cost of the base extrinsic, and with our current configuration the overall weights is increased compared to the old behavior.
Also, the `pallet_transaction_payment` now incurs a cost.

As a result, the weight for the extrinsics is increased in absolute terms, and this increase is more noticeable on lightweight extrinsics, as expected.

I have added the benchmarks and weights for the new extensions, and fixed the test reducing the `FACTOR` for the time being.
Notice, this might break again once we update the weights for all the frame pallets properly, which we will have to do anyway.

For a proper discussion on the subject, also refer to this issue upstream: https://github.com/paritytech/polkadot-sdk/issues/6403

drgora commented on relay-node/cli/src/command.rs:
```diff
@@ -25,6 +25,7 @@ use service::{
 };
 use sp_keyring::Sr25519Keyring;
+#[allow(unused_imports)]
 use std::net::ToSocketAddrs;
```
This is needed for the compilation with the feature `pyroscope`, I gated the import with that feature.
